crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs

```diff
@@ -0,0 +1,137 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::{name::QualifiedName, Arguments, Expr, ExprAttribute, ExprCall};
+use ruff_python_semantic::Modules;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+#[derive(Debug, Eq, PartialEq)]
+enum Replacement {
+    None,
+    Name(String),
+}
+
+/// ## What it does
+/// Checks for uses of deprecated Airflow functions and values.
+///
+/// ## Why is this bad?
+/// Airflow 3.0 removed various deprecated functions, members, and other
+/// values. Some have more modern replacements. Others are considered too niche
+/// and not worth to be maintained in Airflow.
+///
+/// ## Example
+/// ```python
+/// from airflow.utils.dates import days_ago
+///
+///
+/// yesterday = days_ago(today, 1)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from datetime import timedelta
+///
+///
+/// yesterday = today - timedelta(days=1)
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct Airflow3Removal {
+    deprecated: String,
+    replacement: Replacement,
+}
+
+impl Violation for Airflow3Removal {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Airflow3Removal {
+            deprecated,
+            replacement,
+        } = self;
+        match replacement {
+            Replacement::None => format!("`{deprecated}` is removed in Airflow 3.0"),
+            Replacement::Name(name) => {
+                format!("`{deprecated}` is removed in Airflow 3.0; use {name} instead")
+            }
+        }
+    }
+}
+
+struct DeprecatedArgument(Vec<&'static str>, Option<&'static str>);
+
+fn removed_argument(checker: &mut Checker, qualname: &QualifiedName, arguments: &Arguments) {
+    let deprecations = match qualname.segments() {
+        ["airflow", .., "DAG" | "dag"] => vec![DeprecatedArgument(
+            vec!["timetable", "schedule_interval"],
+            Some("schedule"),
+        )],
```

Comment:
I think this is the same as the existing `AIR301`, right? If so, we should remove the `AIR301` implementation.

Reply:
The point of `AIR301` is to detect DAGs without a schedule argument, like this

```python
DAG(dag_id="foo")
```

while the rule here is to detect a DAG using a deprecated schedule argument like

```python
DAG(dag_id="foo", schedule_interval="@daily")
```

We shouldn’t remove AIR301 because AIR302 does not catch the error in the first DAG. I think we should also detect `schedule_interval` and `timetable` in AIR301, so the second DAG only emits AIR302 instead of two errors.

Reply:
Done in the latest commit